pkg/reconciler/taskrun/taskrun.go

```diff
@@ -354,6 +350,12 @@ func (c *Reconciler) prepare(ctx context.Context, tr *v1beta1.TaskRun) (*v1beta1
 		}
 	}
 
+	if verificationResult != nil && verificationResult.VerificationResultType == trustedresources.VerificationError {
+		logger.Errorf("TaskRun %s/%s referred task failed signature verification", tr.Namespace, tr.Name)
+		tr.Status.MarkResourceFailed(podconvert.ReasonResourceVerificationFailed, verificationResult.Err)
+		return nil, nil, controller.NewPermanentError(verificationResult.Err)
+	}
+
```

Comment:
This concerns me a bit that in `default` case of the switch block above, we store the resolved task spec into run status ***before*** checking the verification result. 

I am imaging a case where 
- in the 1st reconcile loop, the verification hasn't finished -> so verificationResult is nil -> we execute the following logics. But note, we have stored the resolved spec into status.
- in the 2nd reconcile loop, the verification finished -> the actual verificationResult should have some error -> but
`GetTaskFuncFromTaskRun` returns a `getTaskfunc` which will return nil `verificationResult` b/c `GetTaskFuncFromTaskRun` sees a spec from the status.

This might surpass the trusted resources verification. WDYT?

Reply:
That's a very interesting thought! If that case happens then yes the verification will be skipped. 
But this is not async code,  I don't know if this will happen -- "the verification hasn't finished but we call the storeTaskSpecAndMergeMeta".

Do you have an example to prove that case?

Reply:
Great point Chuang-- this seems like something that we could add unit tests to cover?

Reply:
We already have tests in reconciler level: https://github.com/tektoncd/pipeline/blob/d4ff4305c2e7b4b56cbb6c781845e98f9a4d7db2/pkg/reconciler/taskrun/taskrun_test.go#L4980 
This test makes sure that at the 1st reconcile loop, we could get the verification error. If the case as @chuangw6  illustrated may happen, then this will become a flaky test right?

Reply:
Can you expand on your concern about why these unit tests might be flaky? You can write unit tests that test results of a reconcile loop that isn't the first one. It sounds like we're looking for a unit test where we previously fetched a resolved spec and stored it in the taskrun status, and now a verification result is available. We'd want to ensure that GetTaskData returns this verification result, rather than a nil verification result. Does this sound like A. something that can happen and B. something that's feasible to add unit tests for?

Reply:
I was saying if Chuang's case is true, then current existing unit tests will be flaky, since we may not get error from trusted resources verification. So I don't think this is possible (again I may be wrong)

Reply:
Discussed with @lbernick offline, to summarize I think this is a good point but I don't think it will happen. 
Though remote resolution seems async, but it is achieved by returning ErrRequestInProgress, https://github.com/tektoncd/pipeline/blob/163ec59f2ccb05dc0aa804de40ac5eb85cd3c7ce/pkg/reconciler/taskrun/taskrun.go#L330-L333 so taskrun will continue to next reconcile loop until we fetch the task. 

But trusted resources verification is currently bundled with remote resolution, the verification will only happen when remote resolution is successful. So in multiple reconcile loops, the remote resolution and verification will be called together in one loop.  And `storeTaskSpecAndMergeMeta` will be called only when remote resolution is done and we return the VerificationResult